### PR TITLE
[FIXED JENKINS-52395] Use the super class of methodType as the invoking class (sender) of the method on the object (receiver)

### PR DESCRIPTION
**Pull request by stuartrowe**
…SuperClass()

See [JENKINS-52395](https://issues.jenkins-ci.org/browse/JENKINS-52395).

### CODE REVIEW

**Comment by stuartrowe** [comment]
@abayer @dwnusbaum @svanoort please review when able. Thanks!

File: lib/src/main/java/com/cloudbees/groovy/cps/sandbox/DefaultInvoker.java [hunk 1]
```diff
@@ -29,7 +29,7 @@ public Object constructorCall(Class lhs, Object[] args) throws Throwable {
     public Object superCall(Class methodType, Object receiver, String method, Object[] args) throws Throwable {
         try {
             MetaClass mc = InvokerHelper.getMetaClass(receiver.getClass());
-            return mc.invokeMethod(methodType, receiver, method, args, true, true);
+            return mc.invokeMethod(methodType.getSuperclass(), receiver, method, args, true, true);
```
Comment (dwnusbaum) [inline lib/src/main/java/com/cloudbees/groovy/cps/sandbox/DefaultInvoker.java]: Any idea if this fixes/affects other issues related to inheritance? For example, this test case for [JENKINS-50731](https://issues.jenkins-ci.org/browse/JENKINS-50731):

```
abstract class SuperClass {
  def someMethod(Map input) {
    privateMethod(input)
  }
  private privateMethod(Map input) {
    input.put("b", 2)
    return input;
  }
}

class SubClass extends SuperClass {
  @Override
  def someMethod(Map input) {
    super.someMethod(input)
  }
}

SubClass x = new SubClass()
Map m = x.someMethod(["a": 1])
echo(m.toString())
```

If you make `SuperClass.privateMethod` public, then this prints `[a:1, b:2]`, but right now it fails with:

> hudson.remoting.ProxyException: groovy.lang.MissingMethodException: No signature of method: SubClass.privateMethod() is applicable for argument types: (java.util.LinkedHashMap) values: [[a:1]]

Comment (stuartrowe) [reply lib/src/main/java/com/cloudbees/groovy/cps/sandbox/DefaultInvoker.java]: I set up a test case for this locally and it doesn't appear to fix that issue:
```
@Test
void transformedSuperClassCallsPrivate() {
    assert evalCPS('''
        class Foo extends CpsTransformerTest.Base {
            private String otherPrivate() {
                return "base"
            }
            public String other() {
                return otherPrivate()
            }
        }
        class Bar extends Foo {
            public String other() {
                return "y"+super.other()
            }
        }
        new Bar().other();
    ''')=="ybase"
}
```

I ran the test with and without the fix and in both cases the test failed with:
```
[ERROR] Errors: 
[ERROR]   CpsTransformerTest.transformedSuperClassCallsPrivate:775->AbstractGroovyCpsTest.evalCPS:49->AbstractGroovyCpsTest.evalCPSonly:55 » MissingMethod
```

Comment (stuartrowe) [reply lib/src/main/java/com/cloudbees/groovy/cps/sandbox/DefaultInvoker.java]: @dwnusbaum I don't have authorization to deploy to the snapshot repository:
```
[ERROR] Failed to execute goal org.sonatype.plugins:nexus-staging-maven-plugin:1.6.5:deploy (injected-nexus-deploy) on project groovy-cps: Failed to deploy artifacts: Could not transfer artifact com.cloudbees:groovy-cps-dgm-builder:jar:1.29-20190620.214529-1 from/to cloudbees-nexus-snapshots (https://oss.sonatype.org/content/repositories/snapshots/): Failed to transfer file: https://oss.sonatype.org/content/repositories/snapshots/com/cloudbees/groovy-cps-dgm-builder/1.29-SNAPSHOT/groovy-cps-dgm-builder-1.29-20190620.214529-1.jar. Return code is: 401, ReasonPhrase: Unauthorized. -> [Help 1]
```
Is this something you could help out with?

Comment (stuartrowe) [reply lib/src/main/java/com/cloudbees/groovy/cps/sandbox/DefaultInvoker.java]: I've sorted out my credentials for https://oss.sonatype.org now which was the original authorization issue. Now I get access denied when trying to deploy:
```
[ERROR] Failed to execute goal org.sonatype.plugins:nexus-staging-maven-plugin:1.6.5:deploy (injected-nexus-deploy) on project groovy-cps: Failed to deploy artifacts: Could not transfer artifact com.cloudbees:groovy-cps-dgm-builder:jar:1.29-20190620.225440-1 from/to cloudbees-nexus-snapshots (https://oss.sonatype.org/content/repositories/snapshots/): Access denied to: https://oss.sonatype.org/content/repositories/snapshots/com/cloudbees/groovy-cps-dgm-builder/1.29-SNAPSHOT/groovy-cps-dgm-builder-1.29-20190620.225440-1.jar, ReasonPhrase: Forbidden. -> [Help 1]
```

Comment (stuartrowe) [reply lib/src/main/java/com/cloudbees/groovy/cps/sandbox/DefaultInvoker.java]: I've created an access request ticket here: https://issues.sonatype.org/browse/OSSRH-49570.

Comment (dwnusbaum) [reply lib/src/main/java/com/cloudbees/groovy/cps/sandbox/DefaultInvoker.java]: Ah, I forgot that this goes to Sonatype and not the normal Jenkins Artifactory where anyone with an account can deploy a snapshot. Nevermind deploying a snapshot for now, we're trying to move this to Artifactory anyway where we would have automatic incremental deployments.

Comment (stuartrowe) [reply lib/src/main/java/com/cloudbees/groovy/cps/sandbox/DefaultInvoker.java]: I'm happy to deploy the snapshot, just need someone with a deployer role from cloudbees to approve/comment on the issue linked above.

Comment (dwnusbaum) [reply lib/src/main/java/com/cloudbees/groovy/cps/sandbox/DefaultInvoker.java]: Yeah, I do not have permission there myself to be able to help.

Comment (stuartrowe) [reply lib/src/main/java/com/cloudbees/groovy/cps/sandbox/DefaultInvoker.java]: @jglick is deploy access to the sonatype repository something you can help with? I deployed a snapshot of groovy-cps to my company's private artifactory repo, but I was unable to test workflow-cps after also pointing it to the same private repo. It appears that Jenkins fails to start properly:

```
[INFO] Running InjectedTest 
[ERROR] Tests run: 7, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 13.231 s <<< FAILURE! - in InjectedTest 
[ERROR] org.jvnet.hudson.test.JellyTestSuiteBuilder$JellyTestSuite(org.jvnet.hudson.test.junit.FailedTest)  Time elapsed: 0.009 s  <<< ERROR!
org.jvnet.hudson.reactor.ReactorException: java.lang.Error: java.lang.reflect.InvocationTargetException
        at org.jvnet.hudson.reactor.Reactor.execute(Reactor.java:282)
        at jenkins.InitReactorRunner.run(InitReactorRunner.java:48)
        at jenkins.model.Jenkins.executeReactor(Jenkins.java:1102)
        at jenkins.model.Jenkins.<init>(Jenkins.java:904)
        at hudson.model.Hudson.<init>(Hudson.java:85)
        at org.jvnet.hudson.test.HudsonTestCase.newHudson(HudsonTestCase.java:486)
        at org.jvnet.hudson.test.HudsonTestCase.setUp(HudsonTestCase.java:317)
        at org.jvnet.hudson.test.JellyTestSuiteBuilder$JellyTestSuite.setUp(JellyTestSuiteBuilder.java:143) 
        at org.jvnet.hudson.test.junit.GroupedTest.run(GroupedTest.java:49)
        at junit.framework.TestSuite.runTest(TestSuite.java:252)
        at junit.framework.TestSuite.run(TestSuite.java:247)
        at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:86)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126) 
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: java.lang.Error: java.lang.reflect.InvocationTargetException
        at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:110)
        at hudson.init.TaskMethodFinder$TaskImpl.run(TaskMethodFinder.java:175)
        at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:296)
        at jenkins.model.Jenkins$5.runTask(Jenkins.java:1068)
        at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:214)
        at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) 
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:104)
        ... 8 more
Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
        at java.util.ArrayList.rangeCheck(ArrayList.java:657)
        at java.util.ArrayList.get(ArrayList.java:433) 
        at hudson.ExtensionList.get(ExtensionList.java:187)
        at org.jenkinsci.plugins.workflow.cps.nodes.StepDescriptorCache.getPublicCache(StepDescriptorCache.java:48)
        at org.jenkinsci.plugins.workflow.cps.nodes.StepDescriptorCache.invalidateGlobalCache(StepDescriptorCache.java:55)
        ... 13 more
```

Comment (jglick) [reply lib/src/main/java/com/cloudbees/groovy/cps/sandbox/DefaultInvoker.java]: > this goes to Sonatype and not the normal Jenkins Artifactory

```sh
mvn clean install source:jar-no-fork deploy:deploy -DaltDeploymentRepository=maven.jenkins-ci.org::default::https://repo.jenkins-ci.org/snapshots/
```

Anyway @dwnusbaum I thought you were going to work on getting this repo moved to @jenkinsci and set up for Incrementals?

Comment (dwnusbaum) [reply lib/src/main/java/com/cloudbees/groovy/cps/sandbox/DefaultInvoker.java]: Yes, KK gave us the go-ahead, we just haven't had a chance to actually move it yet.

Comment (stuartrowe) [reply lib/src/main/java/com/cloudbees/groovy/cps/sandbox/DefaultInvoker.java]: PR against deployed snapshot: https://github.com/jenkinsci/workflow-cps-plugin/pull/300

**Comment by dwnusbaum** [comment]
All tests passed in the downstream PR, so I am going ahead and merging this.

**Comment by stuartrowe** [comment]
@dwnusbaum are you able to provide an ETA for this change to be released?

**Comment by dwnusbaum** [comment]
@stuartrowe Today, if you get a chance to update to groovy-cps 1.29 in https://github.com/jenkinsci/workflow-cps-plugin/pull/300 😄

**Pull request merged**